File: client/packages/client/src/store/privacidadeDoServidor.ts
/**
 * Privacidade POR SERVIDOR.
 *
 * O protocolo Stoat não tem o conceito; ele mora em `UserSettings`, na chave
 * `vortex:privacidadeDoServidor`, sincronizada como as outras preferências.
 * O que cada campo FAZ hoje, e onde:
 *
 * - `filtro` — no cliente: `list/MidiaFiltrada.tsx` soma-o à política de mídia
 *   do servidor. É decisão de quem vê, e é o cliente de quem vê que a aplica.
 * - `dm` e `permitirAmizade` — no `delta`, que LÊ esta chave do destinatário e
 *   recusa conversa nova e pedido de amizade quando o único vínculo entre as
 *   duas pessoas são servidores que dizem não
 *   (`server/crates/core/database/src/util/privacidade_do_servidor.rs`). O
 *   cliente também desvia para a fila de solicitações a DM que já existia.
 * - `mostrarPresenca` — PENDENTE: presença por servidor precisa do `bonfire`
 *   (`pendente/pendencias.ts`, `presencaPorServidor`).
 * - `mostrarAtividade` — sem efeito e fora do modal: o protocolo não tem
 *   atividade de jogo. O campo fica pelo formato já gravado.
 *
 * ⚠ **Por que a decisão é POR SERVIDOR e não uma página de configurações**, e
 * a referência escreve isso: *"é um modal do menu de contexto do servidor,
 * não uma página de configurações — a decisão é sempre tomada no contexto
 * daquele servidor"*. Quem quer fechar DMs de um servidor específico está
 * olhando para ele; abrir configurações e escolher o servidor numa lista
 * inverte a ordem em que a pessoa pensa.
 *
 * `Map` por ID de servidor, e o default é o mesmo objeto para todos — quem
 * nunca mexeu não ocupa entrada. É a razão de `lerPrivacidadeDoServidor`
 * devolver `PADRAO` em vez de criar: sem isso, abrir o modal em cada servidor
 * faria o mapa crescer com cópias idênticas do default.
 */

/** Quem pode me mandar DM a partir deste servidor. */
export const ALCANCES_DE_DM = ["todos", "cargoComum", "ninguem"] as const;
export type AlcanceDeDm = (typeof ALCANCES_DE_DM)[number];

/** O quanto a mídia sensível é escondida. */
export const FILTROS_DE_CONTEUDO = [
  "nao",
  "deNaoAmigos",
  "tudo",
] as const;
export type FiltroDeConteudo = (typeof FILTROS_DE_CONTEUDO)[number];

export type PrivacidadeDoServidor = {
  readonly dm: AlcanceDeDm;
  readonly filtro: FiltroDeConteudo;
  readonly mostrarPresenca: boolean;
  readonly mostrarAtividade: boolean;
  readonly permitirAmizade: boolean;
};

/*
  ⚠ **O padrão é CONSERVADOR, e diverge do design de propósito.** O design
  marca "só quem compartilha cargo" e "filtrar de quem não é amigo". Enquanto
  a preferência não fazia nada, o padrão era só desenho; agora que ela barra DM
  no servidor e vela mídia, esse padrão mudaria o comportamento de quem nunca
  abriu este modal — conversas deixariam de abrir e imagens passariam a chegar
  cobertas, sem ninguém ter escolhido isso. Quem nunca mexeu mantém o de antes:
  DM de qualquer membro, mídia sem filtro pessoal.

  `mostrarAtividade` começa desligado: "o que você está jogando" é dado que
  ninguém pediu para publicar.
*/
export const PADRAO: PrivacidadeDoServidor = {
  dm: "todos",
  filtro: "nao",
  mostrarPresenca: true,
  mostrarAtividade: false,
  permitirAmizade: true,
};

import { avisarSync } from "./sync";

const CHAVE = "vortex:privacidadeDoServidor";

/*
  ⚠ **Versão 2 existe por causa do padrão antigo.** Todo cliente sobe o
  snapshot da chave quando ela falta no servidor, então quem nunca abriu o
  modal JÁ tem gravado `padrao: {dm: "cargoComum", filtro: "deNaoAmigos"}` — o
  padrão de fábrica de antes, não uma escolha. Sem distinguir, o `delta` passaria
  a barrar DM dessas contas. Formato sem versão é lido com a migração abaixo, e
  o servidor o IGNORA inteiro (mesma regra em Rust).
*/
const VERSAO = 2;

function entrada(v: unknown): PrivacidadeDoServidor | undefined {
  if (typeof v !== "object" || v === null) return undefined;
  const r = v as Record<string, unknown>;
  const dm = ALCANCES_DE_DM.find((a) => a === r.dm);
  const filtro = FILTROS_DE_CONTEUDO.find((f) => f === r.filtro);
  if (!dm || !filtro) return undefined;
  if (
    typeof r.mostrarPresenca !== "boolean" ||
    typeof r.mostrarAtividade !== "boolean" ||
    typeof r.permitirAmizade !== "boolean"
  ) {
    return undefined;
  }
  return {
    dm,
    filtro,
    mostrarPresenca: r.mostrarPresenca,
    mostrarAtividade: r.mostrarAtividade,
    permitirAmizade: r.permitirAmizade,
  };
}

/*
  Legado: os dois valores de fábrica antigos viram os conservadores. Quem
  escolheu "cargoComum" de fato perde a escolha — que nunca teve efeito
  nenhum, e por isso não há o que quebrar.
*/
function migrarLegado(e: PrivacidadeDoServidor): PrivacidadeDoServidor {
  return {
    ...e,
    dm: e.dm === "cargoComum" ? "todos" : e.dm,
    filtro: e.filtro === "deNaoAmigos" ? "nao" : e.filtro,
  };
}

function carregarDe(cru: string): {
  padrao: PrivacidadeDoServidor;
  porServidor: Map<string, PrivacidadeDoServidor>;
} {
  try {
    const o: unknown = JSON.parse(cru);
    if (typeof o !== "object" || o === null) {
      return { padrao: PADRAO, porServidor: new Map() };
    }
    const r = o as Record<string, unknown>;
    const atual = r.versao === VERSAO;
    const ajustar = (e: PrivacidadeDoServidor) => (atual ? e : migrarLegado(e));
    const lido = entrada(r.padrao);
    const padraoLido = lido ? ajustar(lido) : PADRAO;
    const mapa = new Map<string, PrivacidadeDoServidor>();
    if (typeof r.porServidor === "object" && r.porServidor !== null) {
      for (const [id, valor] of Object.entries(
        r.porServidor as Record<string, unknown>,
      )) {
        const e = entrada(valor);
        if (e) mapa.set(id, ajustar(e));
      }
    }
    return {
      padrao: igualAoPadrao(padraoLido) ? PADRAO : padraoLido,
      porServidor: mapa,
    };
  } catch {
    return { padrao: PADRAO, porServidor: new Map() };
  }
}

function carregar(): {
  padrao: PrivacidadeDoServidor;
  porServidor: Map<string, PrivacidadeDoServidor>;
} {
  try {
    const cru = localStorage.getItem(CHAVE);
    if (!cru) return { padrao: PADRAO, porServidor: new Map() };
    return carregarDe(cru);
  } catch {
    return { padrao: PADRAO, porServidor: new Map() };
  }
}

const inicial = carregar();
const porServidor = inicial.porServidor;

/*
  O default vigente. Começa no de fábrica e muda só por "aplicar a todos" —
  `PADRAO` continua sendo o de fábrica, para o teste e para a documentação.
*/
let padrao: PrivacidadeDoServidor = inicial.padrao;

/* Referência estável para o default: `lerPrivacidadeDoServidor` depende dela. */
function igualAoPadrao(e: PrivacidadeDoServidor): boolean {
  return (
    e.dm === PADRAO.dm &&
    e.filtro === PADRAO.filtro &&
    e.mostrarPresenca === PADRAO.mostrarPresenca &&
    e.mostrarAtividade === PADRAO.mostrarAtividade &&
    e.permitirAmizade === PADRAO.permitirAmizade
  );
}

export function exportarPrivacidadeDoServidor(): string {
  return JSON.stringify({
    versao: VERSAO,
    padrao,
    porServidor: Object.fromEntries(porServidor),
  });
}

function serializar(): string {
  return exportarPrivacidadeDoServidor();
}

function persistir(): void {
  try {
    localStorage.setItem(CHAVE, serializar());
  } catch {
    /* vale nesta aba */
  }
  avisarSync("vortex:privacidadeDoServidor", serializar());
}

const ouvintes = new Set<() => void>();

export function assinarPrivacidadeDoServidor(ouvinte: () => void): () => void {
  ouvintes.add(ouvinte);
  return () => {
    ouvintes.delete(ouvinte);
  };
}

/**
 * ⚠ Devolve `PADRAO` — a MESMA referência — para quem nunca mexeu.
 *
 * É o que mantém o `getSnapshot` estável, que é a armadilha nº 1 do briefing:
 * montar `{...PADRAO}` aqui daria objeto novo a cada leitura e o
 * `useSyncExternalStore` entraria em laço.
 */
export function lerPrivacidadeDoServidor(
  serverId: string,
): PrivacidadeDoServidor {
  return porServidor.get(serverId) ?? padrao;
}

export function definirPrivacidadeDoServidor(
  serverId: string,
  mudanca: Partial<PrivacidadeDoServidor>,
): void {
  porServidor.set(serverId, {
    ...lerPrivacidadeDoServidor(serverId),
    ...mudanca,
  });
  persistir();
  for (const o of ouvintes) o();
}

/**
 * "Aplicar a todos os servidores".
 *
 * ⚠ **O botão existe porque é o atalho que as pessoas realmente querem depois
 * de configurar o terceiro servidor** — é a frase da referência, e ela
 * descreve o comportamento certo: copiar ESTE servidor para todos os que já
 * têm entrada, e trocar o default para os que não têm. Sem a segunda metade,
 * "todos" significaria "todos os que eu já abri", que é o tipo de promessa
 * parcial que faz alguém desconfiar da tela inteira.
 *
 * Devolve quantos servidores foram afetados, para o toast poder dizer.
 */
export function aplicarATodos(serverId: string): number {
  const escolha = lerPrivacidadeDoServidor(serverId);
  padrao = escolha;
  for (const id of porServidor.keys()) porServidor.set(id, escolha);
  const quantos = porServidor.size;
  persistir();
  for (const o of ouvintes) o();
  return quantos;
}

/** Troca o mapa inteiro — hidratação vinda do servidor. */
export function hidratarPrivacidadeDoServidor(cru: string): void {
  const lido = carregarDe(cru);
  porServidor.clear();
  for (const [id, valor] of lido.porServidor) porServidor.set(id, valor);
  padrao = lido.padrao;
  persistir();
  for (const o of ouvintes) o();
}

/** Estado limpo entre testes. O módulo é global e sobrevive. */
export function limparPrivacidadeDoServidor(): void {
  porServidor.clear();
  padrao = PADRAO;
  try {
    localStorage.removeItem(CHAVE);
  } catch {
    /* mesma regra da escrita */
  }
}

/* ------------------------------------------------------- regras puras */

/**
 * Esta escolha deixa alguém deste servidor abrir DM?
 *
 * `compartilhaCargo` é pergunta que só quem tem os dois membros responde; por
 * isso chega pronta. O mesmo critério roda no `delta`.
 */
export function aceitaDm(
  p: PrivacidadeDoServidor,
  compartilhaCargo: boolean,
): boolean {
  if (p.dm === "todos") return true;
  if (p.dm === "ninguem") return false;
  return compartilhaCargo;
}

/**
 * A privacidade restringe o contato com alguém?
 *
 * ⚠ **Só quando TODOS os servidores em comum dizem não.** Basta um que deixe
 * para haver vínculo legítimo — é o que "vale só aqui, não é global" promete.
 * Sem servidor em comum a preferência por servidor não se aplica, e
 * `undefined` (não deu para saber) não restringe: desviar uma conversa por
 * dado que não chegou seria afirmar algo que ninguém sabe.
 */
export function privacidadeRestringe(
  servidoresEmComum: readonly string[] | undefined,
  permite: (serverId: string) => boolean,
): boolean {
  if (!servidoresEmComum || servidoresEmComum.length === 0) return false;
  return servidoresEmComum.every((id) => !permite(id));
}

/**
 * Alguma escolha restringe DM? Barato, e é o que evita buscar servidores em
 * comum na rede para quem nunca mexeu neste modal.
 */
export function algumaRestricaoDeDm(): boolean {
  if (padrao.dm !== "todos") return true;
  for (const p of porServidor.values()) if (p.dm !== "todos") return true;
  return false;
}

File: server/crates/core/database/src/util/privacidade_do_servidor.rs
//! Vortex: privacidade POR SERVIDOR do destinatário, aplicada a DM e amizade.
//!
//! A preferência não tem modelo próprio: ela mora em `UserSettings`, na chave
//! que o cliente sincroniza (`client/.../store/privacidadeDoServidor.ts`), e o
//! formato aqui é o de lá. As duas pontas precisam concordar; as regras puras
//! deste arquivo repetem as de lá, com testes dos dois lados.
//!
//! Três escolhas que definem o comportamento:
//!
//! - **Só formato versão 2 vale.** Todo cliente sobe o snapshot da chave quando
//!   ela falta, e o padrão de fábrica ANTIGO do cliente era restritivo. Payload
//!   sem versão é de quem nunca escolheu nada, e é ignorado inteiro.
//! - **Chave ausente = nenhuma restrição.** Quem nunca configurou mantém o
//!   comportamento de sempre.
//! - **Só barra quando TODOS os servidores do vínculo dizem não.** Basta um
//!   servidor que deixe, ou um vínculo fora de servidor (amizade, conversa já
//!   existente, grupo em comum), para o contato ser legítimo.

use std::collections::HashMap;

use revolt_result::{create_error, Result};
use serde_json::Value;

use crate::{Database, RelationshipStatus, User};

/// Chave em `UserSettings`.
pub const CHAVE: &str = "vortex:privacidadeDoServidor";

const VERSAO: u64 = 2;

/// Quem pode mandar DM a partir de um servidor.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum AlcanceDeDm {
    Todos,
    CargoComum,
    Ninguem,
}

/// O que importa ao servidor de uma entrada da preferência.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Escolha {
    pub dm: AlcanceDeDm,
    pub permitir_amizade: bool,
}

/// O padrão conservador: igual ao `PADRAO` do cliente.
pub const PADRAO: Escolha = Escolha {
    dm: AlcanceDeDm::Todos,
    permitir_amizade: true,
};

/// A preferência lida.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Preferencias {
    pub padrao: Escolha,
    pub por_servidor: HashMap<String, Escolha>,
}

impl Default for Preferencias {
    fn default() -> Self {
        Preferencias {
            padrao: PADRAO,
            por_servidor: HashMap::new(),
        }
    }
}

impl Preferencias {
    pub fn para(&self, server_id: &str) -> Escolha {
        self.por_servidor
            .get(server_id)
            .copied()
            .unwrap_or(self.padrao)
    }

    /// Alguma escolha pode barrar este tipo de contato? Evita consultas ao
    /// banco para quem nunca restringiu nada.
    pub fn restringe_algo(&self, contato: Contato) -> bool {
        let restritiva = |e: &Escolha| match contato {
            Contato::Dm => e.dm != AlcanceDeDm::Todos,
            Contato::Amizade => !e.permitir_amizade,
        };
        restritiva(&self.padrao) || self.por_servidor.values().any(restritiva)
    }
}

/// Mesma validação do `entrada()` do cliente: entrada incompleta é descartada.
fn entrada(v: &Value) -> Option<Escolha> {
    let o = v.as_object()?;
    let dm = match o.get("dm")?.as_str()? {
        "todos" => AlcanceDeDm::Todos,
        "cargoComum" => AlcanceDeDm::CargoComum,
        "ninguem" => AlcanceDeDm::Ninguem,
        _ => return None,
    };
    let filtro = o.get("filtro")?.as_str()?;
    if !matches!(filtro, "nao" | "deNaoAmigos" | "tudo") {
        return None;
    }
    o.get("mostrarPresenca")?.as_bool()?;
    o.get("mostrarAtividade")?.as_bool()?;
    let permitir_amizade = o.get("permitirAmizade")?.as_bool()?;
    Some(Escolha {
        dm,
        permitir_amizade,
    })
}

/// Lê o texto guardado. Qualquer coisa fora do formato vira "sem restrição".
pub fn ler(cru: &str) -> Preferencias {
    let Ok(v) = serde_json::from_str::<Value>(cru) else {
        return Preferencias::default();
    };
    let Some(o) = v.as_object() else {
        return Preferencias::default();
    };
    if o.get("versao").and_then(Value::as_u64) != Some(VERSAO) {
        return Preferencias::default();
    }
    let padrao = o.get("padrao").and_then(entrada).unwrap_or(PADRAO);
    let por_servidor = o
        .get("porServidor")
        .and_then(Value::as_object)
        .map(|m| {
            m.iter()
                .filter_map(|(id, e)| entrada(e).map(|e| (id.clone(), e)))
                .collect()
        })
        .unwrap_or_default();
    Preferencias {
        padrao,
        por_servidor,
    }
}

/// O tipo de contato sendo iniciado.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Contato {
    Dm,
    Amizade,
}

/// Esta escolha deixa o contato vir deste servidor?
pub fn permite(escolha: Escolha, contato: Contato, compartilham_cargo: bool) -> bool {
    match contato {
        Contato::Amizade => escolha.permitir_amizade,
        Contato::Dm => match escolha.dm {
            AlcanceDeDm::Todos => true,
            AlcanceDeDm::Ninguem => false,
            AlcanceDeDm::CargoComum => compartilham_cargo,
        },
    }
}

/// Barra só quando há servidor no vínculo e todos negam.
pub fn restringe(permissoes: &[bool]) -> bool {
    !permissoes.is_empty() && permissoes.iter().all(|p| !p)
}

/// Recusa o contato de `remetente` com `destinatario` se a privacidade por
/// servidor do destinatário o barrar.
///
/// Erro de LEITURA não barra: a preferência é uma restrição, e falha de banco
/// não é escolha de ninguém.
pub async fn verificar_contato(
    db: &Database,
    remetente: &User,
    destinatario: &User,
    contato: Contato,
) -> Result<()> {
    if remetente.id == destinatario.id
        || remetente.privileged
        || remetente.bot.is_some()
        || destinatario.bot.is_some()
    {
        return Ok(());
    }

    // Vínculos fora de servidor: amizade, ou o destinatário pediu primeiro.
    if matches!(
        remetente.relationship_with(&destinatario.id),
        RelationshipStatus::Friend | RelationshipStatus::Incoming
    ) {
        return Ok(());
    }

    let preferencias = db
        .fetch_user_settings(&destinatario.id, &[CHAVE.to_string()])
        .await
        .ok()
        .and_then(|s| s.get(CHAVE).map(|(_, cru)| ler(cru)))
        .unwrap_or_default();
    if !preferencias.restringe_algo(contato) {
        return Ok(());
    }

    // Conversa já existente ou grupo em comum também são vínculo.
    if db
        .fetch_mutual_channel_ids(&remetente.id, &destinatario.id)
        .await
        .map(|ids| !ids.is_empty())
        .unwrap_or(true)
    {
        return Ok(());
    }

    let servidores: Vec<String> = match contato {
        // Para DM só contam os servidores que liberam DM entre membros: os
        // outros nem dão a permissão, e não são vínculo desta pergunta.
        Contato::Dm => remetente
            .mutual_servers_with_policy(db, &destinatario.id, |s| s.allow_member_dms)
            .await
            .map(|v| v.into_iter().map(|s| s.id).collect())
            .unwrap_or_default(),
        Contato::Amizade => db
            .fetch_mutual_server_ids(&remetente.id, &destinatario.id)
            .await
            .unwrap_or_default(),
    };

    let mut permissoes = Vec::with_capacity(servidores.len());
    for server_id in &servidores {
        let escolha = preferencias.para(server_id);
        let cargo = if contato == Contato::Dm && escolha.dm == AlcanceDeDm::CargoComum {
            compartilham_cargo(db, server_id, &remetente.id, &destinatario.id).await
        } else {
            false
        };
        permissoes.push(permite(escolha, contato, cargo));
    }

    if restringe(&permissoes) {
        Err(create_error!(PrivacyRestricted))
    } else {
        Ok(())
    }
}

async fn compartilham_cargo(db: &Database, server_id: &str, a: &str, b: &str) -> bool {
    let (Ok(ma), Ok(mb)) = (
        db.fetch_member(server_id, a).await,
        db.fetch_member(server_id, b).await,
    ) else {
        return false;
    };
    ma.roles.iter().any(|r| mb.roles.contains(r))
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn escolha(dm: &str, amizade: bool) -> Value {
        json!({
            "dm": dm,
            "filtro": "nao",
            "mostrarPresenca": true,
            "mostrarAtividade": false,
            "permitirAmizade": amizade,
        })
    }

    #[test]
    fn sem_versao_e_ignorado_inteiro() {
        let cru = json!({
            "padrao": escolha("cargoComum", true),
            "porServidor": { "A": escolha("ninguem", false) },
        })
        .to_string();
        assert_eq!(ler(&cru), Preferencias::default());
        assert!(!ler(&cru).restringe_algo(Contato::Dm));
    }

    #[test]
    fn lixo_vira_sem_restricao() {
        for cru in ["", "{", "[]", "42", r#"{"versao":3}"#] {
            assert_eq!(ler(cru), Preferencias::default(), "{cru}");
        }
    }

    #[test]
    fn versao_2_lida_com_entrada_podre_descartada() {
        let cru = json!({
            "versao": 2,
            "padrao": escolha("cargoComum", true),
            "porServidor": {
                "A": escolha("ninguem", false),
                "B": { "dm": "ninguem" },
                "C": escolha("inventado", true),
            },
        })
        .to_string();
        let p = ler(&cru);
        assert_eq!(p.padrao.dm, AlcanceDeDm::CargoComum);
        assert_eq!(
            p.para("A"),
            Escolha {
                dm: AlcanceDeDm::Ninguem,
                permitir_amizade: false
            }
        );
        assert_eq!(p.para("B"), p.padrao);
        assert_eq!(p.para("C"), p.padrao);
        assert!(p.restringe_algo(Contato::Dm));
        assert!(p.restringe_algo(Contato::Amizade));
    }

    #[test]
    fn padrao_ausente_e_conservador() {
        let p = ler(&json!({ "versao": 2 }).to_string());
        assert_eq!(p.padrao, PADRAO);
        assert!(!p.restringe_algo(Contato::Dm));
        assert!(!p.restringe_algo(Contato::Amizade));
    }

    #[test]
    fn permite_dm_e_amizade() {
        let e = |dm, amizade| Escolha {
            dm,
            permitir_amizade: amizade,
        };
        assert!(permite(e(AlcanceDeDm::Todos, false), Contato::Dm, false));
        assert!(!permite(e(AlcanceDeDm::Ninguem, true), Contato::Dm, true));
        assert!(permite(e(AlcanceDeDm::CargoComum, true), Contato::Dm, true));
        assert!(!permite(
            e(AlcanceDeDm::CargoComum, true),
            Contato::Dm,
            false
        ));
        assert!(permite(
            e(AlcanceDeDm::Ninguem, true),
            Contato::Amizade,
            false
        ));
        assert!(!permite(
            e(AlcanceDeDm::Todos, false),
            Contato::Amizade,
            true
        ));
    }

    #[test]
    fn restringe_so_quando_todos_negam() {
        assert!(!restringe(&[]));
        assert!(restringe(&[false]));
        assert!(restringe(&[false, false]));
        assert!(!restringe(&[false, true]));
    }
}
